src/wp-includes/interactivity-api/interactivity-api.php

```diff
@@ -0,0 +1,162 @@
+<?php
+/**
+ * Interactivity API: Functions and hooks
+ *
+ * @package WordPress
+ * @subpackage Interactivity API
+ * @since 6.5.0
+ */
+
+/**
+ * Processes the directives on the rendered HTML of the interactive blocks.
+ *
+ * This processes only one root interactive block at a time because the
+ * rendered HTML of that block contains the rendered HTML of all its inner
+ * blocks, including any interactive block. It does so by ignoring all the
+ * interactive inner blocks until the root interactive block is processed.
+ *
+ * @since 6.5.0
+ *
+ * @param array $parsed_block The parsed block.
+ * @return array The same parsed block.
+ */
+function wp_interactivity_process_directives_of_interactive_blocks( array $parsed_block ): array {
+	static $root_interactive_block = null;
+
+	/*
+	 * Checks whether a root interactive block is already annotated for
+	 * processing, and if it is, it ignores the subsequent ones.
+	 */
+	if ( null === $root_interactive_block ) {
+		$block_name = $parsed_block['blockName'];
+		$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block_name );
+
+		if ( isset( $block_name ) && isset( $block_type->supports['interactivity'] ) && $block_type->supports['interactivity'] ) {
+			// Annotates the root interactive block for processing.
+			$root_interactive_block = array( $block_name, $parsed_block );
+
+			/*
+			 * Adds a filter to process the root interactive block once it has
+			 * finished rendering.
+			 */
+			$process_interactive_blocks = static function ( string $content, array $parsed_block ) use ( &$root_interactive_block, &$process_interactive_blocks ): string {
+				// Checks whether the current block is the root interactive block.
+				list($root_block_name, $root_parsed_block) = $root_interactive_block;
+				if ( $root_block_name === $parsed_block['blockName'] && $parsed_block === $root_parsed_block ) {
+					// The root interactive blocks has finished rendering, process it.
+					$content = wp_interactivity_process_directives( $content );
+					// Removes the filter and reset the root interactive block.
+					remove_filter( 'render_block', $process_interactive_blocks );
+					$root_interactive_block = null;
+				}
+				return $content;
+			};
+
+			/*
+			 * Uses a priority of 20 to ensure that other filters can add additional
+			 * directives before the processing starts.
+			 */
+			add_filter( 'render_block', $process_interactive_blocks, 20, 2 );
```

Comment:
Is the expectation that the directives processing happens as the last thing during rendering?

At the moment, it uses `render_block` filter that is located here:

https://github.com/WordPress/wordpress-develop/blob/a0fedc97f5aab81cbf35f495bc1c0d1b1be245a4/src/wp-includes/class-wp-block.php#L500

There is also another filter a few lines later that could narrow down the handling to the `$root_block_name`:

https://github.com/WordPress/wordpress-develop/blob/a0fedc97f5aab81cbf35f495bc1c0d1b1be245a4/src/wp-includes/class-wp-block.php#L515

Overall, it's not entirely clear why the arbitrary number 20 was selected with this specific filter.

Reply:
It doesn't have to be the last thing, but it should happen after the regular filters just in case those filters add additional directives, which is going to be a common case. 

I'll move it to `render_block_{$this->name}` so it also catches the directives added in that filter.

Reply:
Done in https://github.com/WordPress/wordpress-develop/pull/5953/commits/a2b4aae37dc9dcfc7ecf6a30b9e491f99f5340f9.